File: BOJ/9251/A.py
import sys

sys.stdin = open("D:\Projects\AA\BOJ\9251\input.txt")
input = sys.stdin.readline

A = input().rstrip()
B = input().rstrip()

dp = [[0 for _ in range(1001)] for __ in range(1001)]


def count(a, b):
    cnt = 0
    for i in range(len(a)):
        if a[i] == b[cnt]:
            cnt += 1
            if cnt >= len(b):
                break
    return cnt


j = 0

for i in range(len(A)):
    a = A[i]
    for j in range(len(B)):
        b = B[j]
        add1 = 0
        add2 = 0
        if i-1 >= 0:
            add1 = dp[i-1][j]
        if j-1 >= 0:
            add2 = dp[i][j-1]
        if a == b:
            add3 = 1
            if i-1 >= 0 and j-1 >= 0:
                add3 = dp[i-1][j-1] + 1
            dp[i][j] = add3

        else:
            dp[i][j] = max(add1, add2)


print(dp[len(A)-1][len(B)-1], end="")

File: BOJ/9251/A.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<int> temp;
vector<vector<int>> dp;

int main()
{
    cin.tie(0);
    ios::sync_with_stdio(0);
    freopen("input.txt", "rt", stdin);
    string s1, s2;
    cin >> s1 >> s2;
    temp = vector<int>(s1.size() + 1, 0);
    dp = vector<vector<int>>(s2.size() + 1, temp);
    for (int i = 1; i <= s2.size(); i++)
    {
        char a = s2[i - 1];
        for (int j = 1; j <= s1.size(); j++)
        {
            char b = s1[j - 1];
            if (a == b)
            {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            }
            else
            {
                dp[i][j] = max(dp[i][j - 1], dp[i - 1][j]);
            }
        }
    }
    cout << dp[s2.size()][s1.size()];
}